src/_tkagg.cpp

```diff
@@ -66,61 +80,65 @@ static Tk_PhotoPutBlock_t TK_PHOTO_PUT_BLOCK;
 // extension module or loaded Tcl libraries at run-time.
 static Tcl_SetVar_t TCL_SETVAR;
 
-static PyObject *mpl_tk_blit(PyObject *self, PyObject *args)
+static void
+mpl_tk_blit(py::object interp_obj, const char *photo_name,
+            py::array_t<unsigned char> data, int comp_rule,
+            std::tuple<int, int, int, int> offset, std::tuple<int, int, int, int> bbox)
 {
-    Tcl_Interp *interp;
-    char const *photo_name;
-    int height, width;
-    unsigned char *data_ptr;
-    int comp_rule;
-    int put_retval;
-    int o0, o1, o2, o3;
-    int x1, x2, y1, y2;
+    auto interp = convert_voidptr<Tcl_Interp *>(interp_obj);
+
     Tk_PhotoHandle photo;
-    Tk_PhotoImageBlock block;
-    if (!PyArg_ParseTuple(args, "O&s(iiO&)i(iiii)(iiii):blit",
-                          convert_voidptr, &interp, &photo_name,
-                          &height, &width, convert_voidptr, &data_ptr,
-                          &comp_rule,
-                          &o0, &o1, &o2, &o3,
-                          &x1, &x2, &y1, &y2)) {
-        goto exit;
-    }
     if (!(photo = TK_FIND_PHOTO(interp, photo_name))) {
-        PyErr_SetString(PyExc_ValueError, "Failed to extract Tk_PhotoHandle");
-        goto exit;
+        throw py::value_error("Failed to extract Tk_PhotoHandle");
     }
+
+    auto data_ptr = data.mutable_unchecked<3>();  // Checks ndim and writeable flag.
+    if (data.shape(2) != 4) {
+        throw py::value_error("Data pointer must be RGBA; last dimension is " +
+                              std::to_string(data.shape(2)) + ", not 4");
+    }
+    if (data.shape(0) > INT_MAX) {  // Limited by Tk_PhotoPutBlock argument type.
+        throw std::range_error(
+            "Height (" + std::to_string(data.shape(0)) +
+            ") exceeds maximum allowable size (" + std::to_string(INT_MAX) + ")");
+    }
+    if (data.shape(1) > INT_MAX / 4) {  // Limited by Tk_PhotoImageBlock.pitch field.
+        throw std::range_error(
+            "Width (" + std::to_string(data.shape(1)) +
+            ") exceeds maximum allowable size (" + std::to_string(INT_MAX / 4) + ")");
+    }
+    const auto height = static_cast<int>(data.shape(0));
+    const auto width = static_cast<int>(data.shape(1));
+    const auto
+        x1 = std::get<0>(bbox),
+        x2 = std::get<1>(bbox),
+        y1 = std::get<2>(bbox),
+        y2 = std::get<3>(bbox);
```

Comment:
```suggestion
    int x1, x2, y1, y2;
    std::tie(x1, x2, y1, y2) = bbox;
```
I think use of `tie` is cleaner, but it does need the individual ints to be already defined.

Reply:
I agree it's much cleaner.